test/lib/usd/plugin/mayaGenericWriter.cpp

```diff
@@ -0,0 +1,262 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <mayaUsd/fileio/shaderWriter.h>
+#include <mayaUsd/fileio/shaderWriterRegistry.h>
+#include <mayaUsd/fileio/shading/shadingModeRegistry.h>
+#include <mayaUsd/fileio/utils/adaptor.h>
+#include <mayaUsd/fileio/utils/writeUtil.h>
+#include <mayaUsd/utils/converter.h>
+#include <mayaUsd/utils/util.h>
+
+#include <pxr/base/tf/diagnostic.h>
+#include <pxr/base/tf/staticTokens.h>
+#include <pxr/base/tf/token.h>
+#include <pxr/base/vt/value.h>
+#include <pxr/pxr.h>
+#include <pxr/usd/sdf/valueTypeName.h>
+#include <pxr/usd/usdShade/shader.h>
+#include <pxr/usd/usdShade/tokens.h>
+
+#include <maya/MFnAttribute.h>
+#include <maya/MFnDependencyNode.h>
+#include <maya/MPlug.h>
+#include <maya/MStatus.h>
+#include <maya/MGlobal.h>
+
+#include <basePxrUsdPreviewSurface/usdPreviewSurface.h>
+
+#include <cmath>
+#include <boost/algorithm/string.hpp>
+
+using namespace MAYAUSD_NS;
+
+PXR_NAMESPACE_OPEN_SCOPE
+
+class Maya_GenericWriter : public UsdMayaShaderWriter {
+public:
+    Maya_GenericWriter(
+        const MFnDependencyNode& depNodeFn,
+        const SdfPath&           usdPath,
+        UsdMayaWriteJobContext&  jobCtx);
+
+    void Write(const UsdTimeCode& usdTime) override;
+
+    TfToken GetShadingAttributeNameForMayaAttrName(const TfToken& mayaAttrName) override;
+
+    static ContextSupport CanExport(const UsdMayaJobExportArgs&);
+private:
+    bool _IsConnectable(MPlug & plug);
+};
+
+// Custom registration since all dependency nodes with a drawdb/shader classification are supported
+TF_REGISTRY_FUNCTION_WITH_TAG(UsdMayaShaderWriterRegistry, Maya_GenericWriter)
+{
+    MString nodeTypesCmd("stringArrayToString(listNodeTypes(\"drawdb/shader\"), \" \");");
+    std::string nodeTypes = MGlobal::executeCommandStringResult(nodeTypesCmd).asChar();
+    std::vector<std::string> mayaTypeNames;
+    boost::split(mayaTypeNames, nodeTypes, boost::is_space());
+    mayaTypeNames.push_back("place3dTexture"); // drawdb/geometry
+    for (const std::string& mayaTypeName: mayaTypeNames) {
+        UsdMayaShaderWriterRegistry::Register(
+            TfToken(mayaTypeName),
+            &Maya_GenericWriter::CanExport,
+            [](const MFnDependencyNode& depNodeFn,
+                const SdfPath&           usdPath,
+                UsdMayaWriteJobContext&  jobCtx) {
+                return std::make_shared<Maya_GenericWriter>(depNodeFn, usdPath, jobCtx);
+            });
+    }
+}
+
+TF_DEFINE_PRIVATE_TOKENS(
+    _tokens,
+    // Render context:
+    (maya)
+);
+
+REGISTER_SHADING_MODE_RENDER_CONTEXT(
+    _tokens->maya,
+    "Maya Shaders",
+    "Dumps the bound shader in a Maya UsdShade network that can only be used for import. Will not "
+    "render in the Maya viewport or usdView.");
```

Comment:
The "Will not render" is why this class is in a test module and not in the main plugin. We have no plans to support this quite hackish mode.